packages/grid/src/mouse-handlers/GridColumnMoveMouseHandler.ts

```diff
@@ -1,169 +1,597 @@
+        metrics,
+        true
+      ),
+      draggingColumnDepth,
+      metrics,
+      model
+    );
+
+    const swapColumn = getColumnInfo(
+      GridUtils.getColumnAtX(
+        clamp(
+          isDraggingLeft ? draggingColumn.left : draggingColumn.right,
+          floatingLeftWidth + 1,
+          width
+        ),
+        metrics,
+        true
+      ),
+      draggingColumnDepth,
+      metrics,
+      model
+    );
+
+    if (!swapColumn || !prevSwapColumn) {
+      return;
+    }
+
+    // Cursor has moved past the column drag bounds, don't move the column until we hit the initial offset point again
+    if (this.initialOffset !== this.draggingOffset) {
+      // Pre move < Initial < Post move or vice-versa
+      // User crossed back past the iniital offset point, so we can start moving again
+      if (
+        (this.draggingOffset < this.initialOffset &&
+          this.initialOffset < this.draggingOffset + deltaX) ||
+        (this.draggingOffset > this.initialOffset &&
+          this.initialOffset > this.draggingOffset + deltaX)
+      ) {
+        this.draggingOffset = this.initialOffset;
+        grid.setState({ draggingColumnOffset: this.draggingOffset });
+      } else {
+        // Column can't move since we aren't back at the initial offset yet
+        this.draggingOffset += deltaX;
+        grid.setState({ draggingColumnOffset: this.draggingOffset });
+        return;
+      }
+    }
+
+    // Column can't continue moving in this direction. Pin the left/right side and adjust offset
     if (
-      mouseX < minX &&
-      draggingColumn > 0 &&
-      model.isColumnMovable(grid.getModelColumn(draggingColumn - 1))
+      !model.isColumnMovableTo(
+        draggingColumn.modelIndex,
+        swapColumn.modelIndex,
+        draggingColumnDepth
+      )
     ) {
-      movedColumns = GridUtils.moveItem(
-        draggingColumn,
-        draggingColumn - 1,
-        movedColumns
-      );
-      draggingColumn -= 1;
-    } else if (
-      maxX < mouseX &&
-      draggingColumn < columnCount - 1 &&
-      model.isColumnMovable(grid.getModelColumn(draggingColumn + 1))
+      this.draggingOffset =
+        mouseX - (isDraggingLeft ? swapColumn.right : previousLeft);
+      this.clearScrollInterval();
+      grid.setState({ draggingColumnOffset: this.draggingOffset });
+      return;
+    }
+
+    if (draggingColumn.left <= floatingLeftWidth) {
+      this.setScrollInterval(grid, 'left');
+    } else if (draggingColumn.right > width) {
+      this.setScrollInterval(grid, 'right');
+    } else {
+      this.clearScrollInterval();
+    }
+
+    // Can't swap a column with itself
+    if (swapColumn.visibleIndex === draggingColumn.visibleIndex) {
+      return;
+    }
+
+    const columnMove = this.swapColumns(
+      draggingColumn,
+      swapColumn,
+      movedColumns,
+      model
+    );
+
+    if (!columnMove) {
+      // No swap occurred
+      return;
+    }
+
+    movedColumns = [...movedColumns, columnMove];
```

Comment:
Fixed. Also added a small improvement in `moveItem` and `moveRange` to remove the move if it effectively becomes a no-op. We already ignore when `from === to` from the props, but not when we combined movements. So `1 -> 2, 2 -> 1` would be `1 -> 1` before. Now it will not be included